File: app/cli/user_information.py
"""CLI 侧用户询问事件解析与 resume 构造。"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any


@dataclass(frozen=True, slots=True)
class UserInformationOption:
    """单个可选项。"""

    id: str
    label: str
    value: str = ""


@dataclass(frozen=True, slots=True)
class UserInformationRequest:
    """SSE `user_information_required` 解析结果。"""

    tool_call_id: str
    question: str
    options: list[UserInformationOption]
    allow_multiple: bool
    placeholder: str
    required: bool


@dataclass(frozen=True, slots=True)
class UserInformationAnswer:
    """用户回答，用于构造 resume_value。"""

    tool_call_id: str
    answer: str
    selected_options: list[str]
    cancelled: bool = False

    def to_resume_value(self) -> dict[str, Any]:
        """转为后端 `ResumeUserInformation` 兼容 dict。"""
        return {
            "type": "user_information",
            "tool_call_id": self.tool_call_id,
            "answer": self.answer,
            "selected_options": list(self.selected_options),
            "cancelled": self.cancelled,
        }


class UserInformationCancelled(Exception):
    """用户取消当前信息补充等待；调用方不应 submit resume。"""


def extract_user_information_request(data: dict[str, Any]) -> UserInformationRequest | None:
    """从 SSE `user_information_required.data` 提取询问请求。

    逻辑：
    1. 读取 `user_information_args` 字典；
    2. 解析 `tool_call_id` 与 `question`（`content` 为问题正文兜底）；
    3. 规范化 `options` 列表，过滤无效项。
    """
    args = data.get("user_information_args")
    if not isinstance(args, dict):
        return None
    tool_call_id = str(args.get("tool_call_id") or "").strip()
    question = str(args.get("question") or data.get("content") or "").strip()
    if not tool_call_id or not question:
        return None
    raw_options = args.get("options")
    options: list[UserInformationOption] = []
    if isinstance(raw_options, list):
        for raw in raw_options:
            if not isinstance(raw, dict):
                continue
            option_id = str(raw.get("id") or "").strip()
            label = str(raw.get("label") or "").strip()
            if not option_id or not label:
                continue
            options.append(
                UserInformationOption(
                    id=option_id,
                    label=label,
                    value=str(raw.get("value") or label).strip() or label,
                )
            )
    return UserInformationRequest(
        tool_call_id=tool_call_id,
        question=question,
        options=options,
        allow_multiple=bool(args.get("allow_multiple", False)),
        placeholder=str(args.get("placeholder") or ""),
        required=bool(args.get("required", True)),
    )


def build_answer_from_text(request: UserInformationRequest, text: str) -> UserInformationAnswer:
    """构造自由文本回答。"""
    return UserInformationAnswer(
        tool_call_id=request.tool_call_id,
        answer=str(text or "").strip(),
        selected_options=[],
    )


def build_answer_from_options(
    request: UserInformationRequest,
    selected_ids: list[str],
) -> UserInformationAnswer:
    """构造选项式回答，并生成可读 `answer` 文本。"""
    selected_set = {str(item).strip() for item in selected_ids if str(item).strip()}
    labels: list[str] = []
    for item in request.options:
        if item.id in selected_set:
            labels.append(item.label)
    answer = ", ".join(labels)
    return UserInformationAnswer(
        tool_call_id=request.tool_call_id,
        answer=answer,
        selected_options=sorted(selected_set),
    )

File: client/internal/hitl/user_information.go
package hitl

import (
	"fmt"
	"io"
	"strings"
)

// UserInformationOption 为 ask_user_information 的单个可选项。
type UserInformationOption struct {
	ID    string
	Label string
	Value string
}

// UserInformationRequest 为 SSE user_information_required 解析结果。
type UserInformationRequest struct {
	ToolCallID    string
	Question      string
	Options       []UserInformationOption
	AllowMultiple bool
	Placeholder   string
	Required      bool
}

// ExtractUserInformationRequest 从 SSE data 提取询问请求；无效时返回 nil。
func ExtractUserInformationRequest(data map[string]any) *UserInformationRequest {
	if data == nil {
		return nil
	}
	args, _ := data["user_information_args"].(map[string]any)
	if args == nil {
		return nil
	}
	toolCallID := strings.TrimSpace(fmt.Sprint(args["tool_call_id"]))
	question := strings.TrimSpace(fmt.Sprint(args["question"]))
	if question == "" {
		question = strings.TrimSpace(fmt.Sprint(data["content"]))
	}
	if toolCallID == "" || question == "" {
		return nil
	}
	req := &UserInformationRequest{
		ToolCallID:    toolCallID,
		Question:      question,
		AllowMultiple: boolArg(args["allow_multiple"]),
		Placeholder:   strings.TrimSpace(fmt.Sprint(args["placeholder"])),
		Required:      boolArgDefault(args["required"], true),
	}
	rawOptions, _ := args["options"].([]any)
	for _, raw := range rawOptions {
		m, ok := raw.(map[string]any)
		if !ok {
			continue
		}
		id := strings.TrimSpace(fmt.Sprint(m["id"]))
		label := strings.TrimSpace(fmt.Sprint(m["label"]))
		if id == "" || label == "" {
			continue
		}
		value := strings.TrimSpace(fmt.Sprint(m["value"]))
		if value == "" {
			value = label
		}
		req.Options = append(req.Options, UserInformationOption{ID: id, Label: label, Value: value})
	}
	return req
}

// BuildUserInformationResume 构造 user_information resume_value。
func BuildUserInformationResume(req *UserInformationRequest, answer string, selectedIDs []string, cancelled bool) map[string]any {
	if req == nil {
		return map[string]any{"type": "user_information"}
	}
	rv := map[string]any{
		"type":         "user_information",
		"tool_call_id": req.ToolCallID,
		"answer":       strings.TrimSpace(answer),
		"selected_options": append([]string(nil), selectedIDs...),
	}
	if cancelled {
		rv["cancelled"] = true
	}
	return rv
}

// BuildUserInformationResumeFromOptions 根据选中 option id 构造 resume（answer 为 label 拼接）。
func BuildUserInformationResumeFromOptions(req *UserInformationRequest, selected map[string]bool) (map[string]any, error) {
	if req == nil {
		return nil, fmt.Errorf("empty request")
	}
	selectedIDs := make([]string, 0)
	labels := make([]string, 0)
	for _, opt := range req.Options {
		if selected[opt.ID] {
			selectedIDs = append(selectedIDs, opt.ID)
			labels = append(labels, opt.Label)
		}
	}
	if len(selectedIDs) == 0 && req.Required {
		return nil, fmt.Errorf("请至少选择一项")
	}
	return BuildUserInformationResume(req, strings.Join(labels, ", "), selectedIDs, false), nil
}

// PrintUserInformationTranscript 向 w 输出合并后的「Agent 询问」块（REPL stdin 路径）。
func PrintUserInformationTranscript(w io.Writer, req *UserInformationRequest) {
	for _, line := range FormatUserInformationTranscriptLines(req) {
		fmt.Fprintln(w, line)
	}
}

// FormatUserInformationTranscriptLines 将询问合并为 transcript 单条（对齐 Python TUI「Agent 询问」块）。
func FormatUserInformationTranscriptLines(req *UserInformationRequest) []string {
	if req == nil || strings.TrimSpace(req.Question) == "" {
		return nil
	}
	lines := []string{"[tool] ● Agent 询问"}
	for _, part := range strings.Split(req.Question, "\n") {
		lines = append(lines, "    "+strings.TrimRight(part, " \t"))
	}
	return lines
}

// FormatUserInformationOptions 格式化选项列表（供 TUI 展示）。
func FormatUserInformationOptions(req *UserInformationRequest, selected map[string]bool, cursor int) string {
	if req == nil || len(req.Options) == 0 {
		return ""
	}
	var b strings.Builder
	for i, opt := range req.Options {
		mark := "[ ]"
		if selected[opt.ID] {
			mark = "[x]"
		}
		prefix := "  "
		if i == cursor {
			prefix = "> "
		}
		fmt.Fprintf(&b, "%s%s %s\n", prefix, mark, opt.Label)
	}
	hint := "↑/↓ 移动 · Space 切换 · Enter 确认"
	if req.AllowMultiple {
		hint += "（可多选）"
	} else {
		hint += "（单选）"
	}
	fmt.Fprintf(&b, "\n%s", hint)
	return strings.TrimRight(b.String(), "\n")
}

func boolArg(v any) bool {
	switch t := v.(type) {
	case bool:
		return t
	case string:
		return strings.EqualFold(t, "true") || t == "1"
	default:
		return false
	}
}

func boolArgDefault(v any, def bool) bool {
	if v == nil {
		return def
	}
	return boolArg(v)
}
